wgengine/magicsock/magicsock_linux.go

```diff
@@ -213,6 +213,19 @@ func (c *Conn) receiveDisco(pc net.PacketConn) {
 			// Too small to be a valid UDP datagram, drop.
 			continue
 		}
+
+		dstPort := binary.BigEndian.Uint16(buf[2:4])
+
+		// TODO(raggi): unsure if these are memoized by pcon, if not we should do so.
+		v4Port := c.pconn4.LocalAddr().Port
```

Comment:
You get correct port numbers back from this? Our original change had almost exactly this, but the pconns returned Port=0 all the time, leading to dropping all disco traffic.

Reply:
I'm sure on a system without ipv6 for example, the v6 one returns 0, as it has a zero value explicitly returned from the fake conn that's in there.

A 0 compare should not result in all traffic being dropped however, as it shouldn't match the incoming port number. I can explicitly skip 0 values as a defensive measure though.

Were you getting 0 ports from the UDP header? That would be more concerning as it could lead to a broader issue.

Reply:
No, the port numbers in the packet were fine, but all disco packets were getting dropped because the comparisons were all 0 == 39482 or whatever.

Are you running tailscaled with a --port argument? I was running in userspace mode without --port, if that's relevant to what I saw.

It's also possible my code was just wrong at the time, I didn't spend much time trying to debug it because I thought it was unnecessary.

Reply:
Ah, userspace mode, interesting. I shall have to test that.

Reply:
Userspace seems fine. I also tested with both a fixed port number (which effectively mitigates the original issue too), and with `--port 0`.

I'm not able to reproduce the condition you describe.